x/ibc/04-channel/manager.go

```diff
@@ -0,0 +1,306 @@
+		return Object{}, errors.New("invalid channel id")
+	}
+
+	if channel.Port != man.port {
+		return Object{}, errors.New("invalid port")
+	}
+
+	return man.man.Create(ctx, portid, chanid, channel)
+}
+
+func (man PortManager) Query(ctx sdk.Context, portid, chanid string) (Object, error) {
+	if !man.chanid(chanid) {
+		return Object{}, errors.New("invalid channel id")
+	}
+
+	obj, err := man.man.Query(ctx, portid, chanid)
+	if err != nil {
+		return Object{}, err
+	}
+
+	if obj.Value(ctx).Port != man.port {
+		return Object{}, errors.New("invalid port")
+	}
+
+	return obj, nil
+}
+*/
+
+type Object struct {
+	chanid string
+	portid string
+
+	Channel state.Value
+
+	SeqSend state.Integer
+	SeqRecv state.Integer
+	Packets state.Indexer
+
+	Available state.Boolean
+
+	Connections []connection.Object
+
+	counterparty CounterObject
+}
+
+type CounterObject struct {
+	chanid string
+	portid string
+
+	Channel commitment.Value
+
+	SeqSend commitment.Integer
+	SeqRecv commitment.Integer
+	Packets commitment.Indexer
+
+	Available commitment.Boolean
+
+	Connections []connection.CounterObject
+}
+
+func (obj Object) OriginConnection() connection.Object {
+	return obj.Connections[0]
+}
+
+func (obj Object) Context(ctx sdk.Context, proofs []commitment.Proof) (sdk.Context, error) {
+	return obj.OriginConnection().Context(ctx, nil, proofs)
+}
+
+func (obj Object) ChanID() string {
+	return obj.chanid
+}
+
+func (obj Object) GetChannel(ctx sdk.Context) (res Channel) {
+	obj.Channel.Get(ctx, &res)
+	return
+}
+
+func (obj Object) PacketCommit(ctx sdk.Context, index uint64) []byte {
+	return obj.Packets.Value(index).GetRaw(ctx)
+}
+
+/*
+func (obj Object) Sendable(ctx sdk.Context) bool {
+	return obj.connection
+}
+
+func (obj Object) Receivable(ctx sdk.Context) bool {
+	return kinds[obj.kind.Get(ctx)].Receivable
+}
+*/
+func (obj Object) exists(ctx sdk.Context) bool {
+	return obj.Channel.Exists(ctx)
+}
+
+func (man Manager) Send(ctx sdk.Context, portid, chanid string, packet Packet) error {
+	/*
+		if !obj.Sendable(ctx) {
+			return errors.New("cannot send Packets on this channel")
+		}
+	*/
+	obj, err := man.Query(ctx, portid, chanid)
+	if err != nil {
+		return err
+	}
+
+	if obj.OriginConnection().Client.GetConsensusState(ctx).GetHeight() >= packet.Timeout() {
+		return errors.New("timeout height higher than the latest known")
+	}
+
+	obj.Packets.Set(ctx, obj.SeqSend.Increment(ctx), packet)
+
+	return nil
+}
+
+func (man Manager) Receive(ctx sdk.Context, proofs []commitment.Proof, portid, chanid string, packet Packet) error {
+	obj, err := man.Query(ctx, portid, chanid)
+	if err != nil {
+		return err
+	}
+
+	/*
+		if !obj.Receivable(ctx) {
+			return errors.New("cannot receive Packets on this channel")
+		}
+	*/
+
+	ctx, err = obj.Context(ctx, proofs)
+	if err != nil {
+		return err
+	}
+
+	err = assertTimeout(ctx, packet.Timeout())
+	if err != nil {
+		return err
+	}
+
+	// XXX: increment should happen before verification, reflect on the spec
```

Comment:
This doesn't look like it performs most of the requisite checks, and it also doesn't look like it supports unordered channels.

To confirm, this is supposed to be `recvPacket` from the spec, right?